kernel/src/process/clone.rs

```diff
@@ -671,25 +671,93 @@ fn create_child_process(
     child_proc
 }
 
-fn set_parent_and_group(parent: &Process, child: &Arc<Process>) {
-    // Lock order: children of process -> process table -> group of process
-    // -> group inner -> session inner
-    let mut children_mut = parent.children().lock();
+fn set_parent_and_group(clone_flags: CloneFlags, parent: &Arc<Process>, child: &Arc<Process>) {
+    loop {
+        let real_parent = clone_parent(clone_flags, parent);
+
+        // Lock the parent's children before checking its status.
+        let mut children_mut = real_parent.children().lock();
+
+        if real_parent.status().is_zombie() {
+            // The real parent is concurrently exiting group.
+            // Wait until it finishes exiting and then retry.
+            drop(children_mut);
+            drop(real_parent);
+            Thread::yield_now();
```

Comment:
Consider adding a flag to `children: Mutex<BtreeMap>` or consider something like `children Mutex<Option<BTreeMap>>`? `Thread::yield_now` sounds a bit odd.